File: java/decouple-monolith/exercise/src/main/java/compliance/temporal/ComplianceNexusServiceImpl.java
package compliance.temporal;

// ═══════════════════════════════════════════════════════════════════
//  TODO 2: Implement the Nexus service handlers
// ═══════════════════════════════════════════════════════════════════
//
// This class implements the ComplianceNexusService interface.
// It has two handlers — one async, one sync:
//
//   1. Add @ServiceImpl(service = ComplianceNexusService.class) to the class
//   2. Annotate BOTH methods with @OperationImpl
//
// ── checkCompliance (async — fromWorkflowHandle) ────────────────────
//
// This handler receives compliance check requests from the Payments team and
// starts a ComplianceWorkflow to process them.
//
// Use fromWorkflowHandle — NOT OperationHandler.sync()
//     fromWorkflowHandle ensures retries reuse the same workflow instead of
//     creating duplicate workflows. This is the correct pattern for long-running work.
//
//   return WorkflowRunOperation.fromWorkflowHandle((ctx, details, input) -> {
//       WorkflowClient client = Nexus.getOperationContext().getWorkflowClient();
//       ComplianceWorkflow wf = client.newWorkflowStub(
//               ComplianceWorkflow.class,
//               WorkflowOptions.newBuilder()
//                       .setTaskQueue("compliance-risk")
//                       .setWorkflowId("compliance-" + input.getTransactionId())
//                       .build());
//       return WorkflowHandle.fromWorkflowMethod(wf::run, input);
//   });
//
// ── submitReview (sync — OperationHandler.sync) ─────────────────────
//
// This handler receives a review decision (approve/deny) and forwards it to the
// already-running ComplianceWorkflow as a Workflow Update.
//
// Use OperationHandler.sync because the Update returns immediately (the workflow
// just stores the result). This completes well within the 10-second handler deadline.
//
// Key difference:
//   - checkCompliance uses fromWorkflowHandle because it STARTS a new long-running workflow
//   - submitReview uses OperationHandler.sync because it INTERACTS with an existing workflow
//
//   return OperationHandler.sync((ctx, details, input) -> {
//       WorkflowClient client = Nexus.getOperationContext().getWorkflowClient();
//       ComplianceWorkflow wf = client.newWorkflowStub(
//               ComplianceWorkflow.class,
//               "compliance-" + input.getTransactionId());
//       return wf.review(input.isApproved(), input.getExplanation());
//   });

import compliance.domain.ComplianceRequest;
import compliance.domain.ComplianceResult;
import compliance.temporal.workflow.ComplianceWorkflow;
import io.nexusrpc.handler.OperationHandler;
import io.nexusrpc.handler.OperationImpl;
import io.nexusrpc.handler.ServiceImpl;
import io.temporal.client.WorkflowClient;
import io.temporal.client.WorkflowOptions;
import io.temporal.nexus.Nexus;
import io.temporal.nexus.WorkflowRunOperation;
import io.temporal.nexus.WorkflowHandle;
import shared.domain.ReviewRequest;
import shared.nexus.ComplianceNexusService;

// TODO 2: Add @ServiceImpl(service = ComplianceNexusService.class) annotation
public class ComplianceNexusServiceImpl {

    // TODO 2: Add @OperationImpl and implement checkCompliance() using fromWorkflowHandle
    public OperationHandler<ComplianceRequest, ComplianceResult> checkCompliance() {
        return null;
    }

    // TODO 2: Add @OperationImpl and implement submitReview() using OperationHandler.sync
    public OperationHandler<ReviewRequest, ComplianceResult> submitReview() {
        return null;
    }
}

File: kotlin/decouple-monolith/exercise/src/main/kotlin/compliance/temporal/ComplianceNexusServiceImpl.kt
package compliance.temporal

import compliance.domain.ComplianceRequest
import compliance.domain.ComplianceResult
import compliance.temporal.workflow.ComplianceWorkflow
import io.nexusrpc.handler.OperationHandler
import io.temporal.client.WorkflowOptions
import io.temporal.nexus.Nexus
import io.temporal.nexus.WorkflowHandle
import io.temporal.nexus.WorkflowRunOperation
import shared.domain.ReviewRequest
import shared.nexus.ComplianceNexusService

/**
 * The Compliance team's Nexus handler. Payments depends on the contract in
 * ComplianceNexusService. This file is the code that answers those calls, and only
 * the Compliance team owns it.
 *
 * This is the hardest file in the lab. Two things you need in both methods:
 *
 *   Nexus.getOperationContext().workflowClient
 *     A Temporal Client, already connected to this team's Namespace.
 *
 *   "compliance-{transactionId}"
 *     The Workflow ID to use. Same ID both times: one method creates that Workflow,
 *     the other looks up the same one. The Task Queue is "compliance-risk".
 *
 * Stuck? "Ask AI" on https://docs.temporal.io:
 *   "how do I start a Workflow from a Nexus Operation handler in Java?"
 * Remember to search the Java docs, not Kotlin.
 */
// ── TODO 4 ──────────────────────────────────────────────────────────────────────
// Annotate this class so Temporal knows it answers calls for the contract. The
// annotation has to name the interface this class implements.
class ComplianceNexusServiceImpl {

    // ── TODO 5 ──────────────────────────────────────────────────────────────────
    // Two things here: annotate this method as the code that runs when Payments calls
    // checkCompliance, then write its body. Miss the annotation and the Worker refuses
    // to start with "Missing handlers for service operations".
    //
    // This is the SLOW call. A risky payment has to be approved by a person, so an
    // answer can be minutes or hours away. This method must not sit and wait for it.
    //
    // Instead: start a ComplianceWorkflow and return immediately, handing back a
    // reference to it. Temporal holds onto that reference and collects the result
    // whenever the Workflow finishes. Use WorkflowRunOperation for this.
    //
    // Do NOT use OperationHandler.sync here. That one waits for the answer, and it
    // gets cut off after 10 seconds. Worse, when Temporal retries the failed call it
    // would start a SECOND compliance check for the same payment.
    //
    // You need two pieces:
    //   1. a ComplianceWorkflow stub, built with WorkflowOptions that set the Task
    //      Queue and the Workflow ID
    //   2. a WorkflowHandle built from that stub's Workflow method
    // The imports at the top are exactly the ones you need, and nothing more.
    fun checkCompliance(): OperationHandler<ComplianceRequest, ComplianceResult> {
        TODO("TODO 5: start a ComplianceWorkflow and return a reference to it")
    }

    // ── TODO 6 ──────────────────────────────────────────────────────────────────
    // Mark this method as the code that runs when Payments calls submitReview, then
    // fill it in.
    //
    // This is the FAST call. By now the compliance check is already running and is
    // parked waiting for a decision. All you do is find that Workflow and hand it the
    // yes or no. That takes milliseconds, so this method can wait for the answer and
    // return it in the same call. Use OperationHandler.sync.
    //
    // Look the Workflow up by its ID, then call review() on it.
    //
    // One Kotlin wrinkle: OperationHandler.sync says its input can be null, so Kotlin
    // sees ReviewRequest? and you need a non-null assertion (!!) before using it.
    fun submitReview(): OperationHandler<ReviewRequest, ComplianceResult> {
        TODO("TODO 6: find the running Workflow and give it the review decision")
    }
}
